### PR TITLE
use a deep find to work out submodules vs use a file

### PR DESCRIPTION
**Pull request by michaelneale**
@charleso FYI here is a version of the change using find

### CODE REVIEW

File: src/cabal [hunk 1]
```diff
@@ -60,24 +60,24 @@ initialize () {
...
-        )
+    for submodule in $(find lib -maxdepth 4 ! -path "lib/*/bin/*" -name \*.cabal | sed 's/[^\/]*$//g'| sed 's/\/$//g');
+    do
+        if [ -d "$submodule" ]; then
+           find "$submodule" -maxdepth 2 -name \*.cabal
+        fi
+    done | grep -q . || (
+        cd $(git rev-parse --show-toplevel)
+        git submodule init
+        git submodule update
```
Comment (charleso) [inline src/cabal]: While we're here, this can be `git submodule update --init`

File: src/cabal [hunk 2]
```diff
@@ -60,24 +60,24 @@ initialize () {
...
-        cat $project.submodules | while read submodule; do
-            if [ -d "$submodule" ]; then
-               find "$submodule" -maxdepth 2 -name \*.cabal
-            fi
-        done | grep -q . || (
-            cd $(git rev-parse --show-toplevel)
-            git submodule init
-            git submodule update
-        )
+    for submodule in $(find lib -maxdepth 4 ! -path "lib/*/bin/*" -name \*.cabal | sed 's/[^\/]*$//g'| sed 's/\/$//g');
```
Comment (charleso) [inline src/cabal]: Can/should we use `git submodule` and grep for lines that start with `-`?

**Comment by charleso** [comment]
Nice one! :+1:

File: src/cabal (hunk 2 above)
Comment (charleso) [inline src/cabal]: > git submodule | grep "^-" | cut -d ' ' -f 2

(Maybe)

**Comment by fractalcat** [comment]
LGTM.

**Comment by charleso** [comment]
@michaelneale Maybe we should add a deprecation check that lets the user know they can delete the `$project.submodules` file. :)

File: src/cabal (hunk 2 above)
Comment (michaelneale) [inline src/cabal]: yeah sounds good.

File: src/cabal [hunk 3]
```diff
@@ -60,24 +60,24 @@ initialize () {
...
+        if [ -d "$submodule" ]; then
+           find "$submodule" -maxdepth 2 -name \*.cabal
+        fi
+    done | grep -q . || (
+        cd $(git rev-parse --show-toplevel)
+        git submodule init
+        git submodule update
+    )
+
+    for submodule in $(find lib -maxdepth 4 ! -path "lib/*/bin/*" -name \*.cabal | sed 's/[^\/]*$//g'| sed 's/\/$//g');
```
Comment (fractalcat) [inline src/cabal]: Why not `find lib -maxdepth 4 ! -path "lib/*/bin/*" -name \*.cabal | sed 's@/[^\/]*$@@g'`?

**Comment by etorreborre** [comment]
Great job @michaelneale :santa:!

File: src/cabal (hunk 3 above)
Comment (michaelneale) [inline src/cabal]: thanks @fractalcat I knew you would come up with something - perfect.

**Comment by nhibberd** [comment]
Any reason for only maxdepth 4?

File: src/cabal (hunk 3 above)
Comment (markhibberd) [inline src/cabal]: Isn't  `sed 's@/[^/]*$@@g'` just `xargs -L 1 dirname`? Not that it matters, just checking I am reading it right.

Comment (fractalcat) [inline src/cabal]: Yes, now that you mention it. `dirname` is probably clearer than the regex.

Comment (michaelneale) [inline src/cabal]: yes, no regex means one less problem. Not sure why I didn't think of dirname. Perhaps as the curator or @EvilBashTips twitter account I feel repressed for using just shell and not bash subconsciously.

**Comment by michaelneale** [comment]
@nhibberd it seemed to be needed for some of our libs - but are there any that have deeper cabal files? 
depth of 4 not enough?

**Comment by nhibberd** [comment]
I would think they're would be more in some top level projects, not sure though. Thinking about ordinance/hydra etc

**Comment by michaelneale** [comment]
@nhibberd @markhibberd oh this may be a show stopper: https://github.com/ambiata/hydra/tree/master/hydra-core

the lib is at the top level, not in the hydra-core subdir - so a find won't work unless you know where the root is that container the lib? (whereas the submodules file tells it?)

Any ideas?

**Comment by fractalcat** [comment]
Seems like a symlink to `../lib/` in `hydra-core` would work, though it's not a pretty solution.

**Comment by michaelneale** [comment]
@fractalcat yeah and it breaks the "just works" thing. I could make it so it looks for a .submodules file, uses it, but if if doesn't, uses the find? @charleso what you think?  Don't like changes to extras that break existing things, has to be seamless upgrade.

**Comment by markhibberd** [comment]
@michaelneale It depends on how specific to our code structure you go, but I all the "multi module" projects I know of look like that (hydra, x), so I think it is safe to run the find from `$(git rev-parse --show-toplevel)` rather than where the cabal file is. But it does mean it relies on a pretty specific convention which isn't great.

**Comment by michaelneale** [comment]
@markhibberd yeah - I think that would work too - are there cases where there are lib/submodules deeper down in the tree? Its a reasonable convention (or defaulting to looking for a $project.submodules file could work too?

**Comment by markhibberd** [comment]
There maybe cases like that given the number of different keyboards doing it, but I think top-level always makes the most sense. If you have a single repo there isn't a really good reason to have inconsistent dependencies between modules so having a single set makes sense. I don't mind the default to looking at submodules file as well, and perhaps having it do both gives us an escape hatch for migrating things that we may need anyway.

**Comment by michaelneale** [comment]
@fractalcat @charleso @markhibberd made some changes, sorry took a while as wanted to test it with a few empty projects of different types. 

This will now use the $project.submodules file if it is there, if not, it will discover them from the lib in the root of the project (giving people a choice on if they use it). If it is ok, can update the templates to use the latest script, as well as drop their .submodules files to encourage future non-use of them (just one less thing).

File: src/cabal [hunk 4]
```diff
@@ -60,24 +60,36 @@ initialize () {
...
+           find "$submodule" -maxdepth 2 -name \*.cabal
+        fi
+    done | grep -q . || (
+        cd $PROJECT_ROOT
+        git submodule update --init
+    )
+
+    # discover and add cabal sources
+    CABAL_SOURCES=$(cd "$PROJECT_ROOT" && find lib -maxdepth 4 ! -path "lib/*/bin/*" -name \*.cabal | xargs -L 1 dirname)
+    [ ! -f "$project.submodules" ] || {
```
Comment (charleso) [inline src/cabal]: Worth making this an `if` to avoid the hit if the submodules file exists?

**Comment by charleso** [comment]
:+1:

**Comment by fractalcat** [comment]
:+1:

**Comment by michaelneale** [comment]
@markhibberd how this look?

File: src/cabal (hunk 4 above)
Comment (michaelneale) [inline src/cabal]: adds bulk for no real speed up (the find is negligible)

**Comment by michaelneale** [comment]
bump - people ok with this?

**Comment by etorreborre** [comment]
:+1: go for it.

**Comment by markhibberd** [comment]
:large_blue_circle:

**Pull request merged**